### ISSUE TITLE
Still got CORS headers when enabling CORS for API Gateway and sending get request without Origin header

### ISSUE THREAD

**Issue by cherrycl** [edgexfoundry/edgex-go#4648]
### Description and Minimal Reproduction [**REQUIRED**]
1. Get compose file from [edgex-compose/docker-compose.yml](https://github.com/edgexfoundry/edgex-compose/blob/main/docker-compose.yml)
2. Add environment variable `EDGEX_SERVICE_CORSCONFIGURATION_ENABLECORS: true` under `security-proxy-setup` service
3. Deploy EdgeX by the above compose file
4. Send get request without Origin header
`curl -v -k -H "Authorization: Bearer xxxx" https://localhost:8443/core-metadata/api/v3/ping `

Response headers:
```
...
< HTTP/1.1 200 OK
< Server: nginx
< Date: Wed, 23 Aug 2023 03:12:43 GMT
< Content-Type: application/json
< Content-Length: 92
< Connection: keep-alive
< X-Correlation-Id: 
< Access-Control-Allow-Origin: https://localhost
< Access-Control-Expose-Headers: Cache-Control, Content-Language, Content-Length, Content-Type, Expires, Last-Modified, Pragma, X-Correlation-ID
< Vary: origin
...
```

The response still contains CORS header. This behavior for API Gateway is correct or not?

## 🌍  Your Environment
**Deployment Environment:** Ubuntu 22.04 / x86

**EdgeX Version [**REQUIRED**]:** Main

**Anything else relevant?**

**Comment by cloudxxx8** [comment]
@bnevis-i this issue is just a question to confirm the correct behavior.  we have implemented the CORS test cases for the individual micro service.  Do you have any suggestion about what we should test for API Gateway behaviors?

**Comment by bnevis-i** [comment]
> @bnevis-i this issue is just a question to confirm the correct behavior. we have implemented the CORS test cases for the individual micro service. Do you have any suggestion about what we should test for API Gateway behaviors?

You are right that this is a bit off.  NGINX doesn't allow nested IF statements or complex conditionals (AND/OR) etc.  I have to rewrite the logic to only send out CORS headers IFF `Origin:` is set, and only send the preflight headers IFF `Origin:` is set AND the method is OPTIONS.  There are no complex booleans allowed in NGINX `if` statements so the second case is tricky.

I think it is a valid finding, but first and foremost we need to make sure that CORS requests actually work.  Avoiding sending "extra" headers is a secondary concern.

**Comment by bnevis-i** [comment]
@cherrycl Based on https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS documentation, I have made the following truth table:

**Operation**|**Preflight Request**|**CORS Request Header**|**if http_Origin**|**if http_origin && OPTIONS**|**if http_origin && !OPTIONS**
:-----:|:-----:|:-----:|:-----:|:-----:|:-----:
Access-Control-Allow-Origin|x|x|x| | 
Access-Control-Allow-Methods|x| | |x| 
Access-Control-Allow-Headers|x| | |x| 
Access-Control-Allow-Allow-Credentials|x|x|x| | 
Access-Control-Allow-Max-Age|x| | |x| 
Access-Control-Expose-Headers| |x| | |x
Vary|x|x|x| | 

The first two columns signify if the header is returned in response to a preflight request or an actual CORS request.  The last three columns signify the request header logic that should be used for sending the response.

It is not possible to write compound IF conditions in NGINX or nested IF conditions, but it is possible to be clever about it by building a string, as detailed in <https://ezecodes.wordpress.com/2016/06/30/multiple-if-conditions-in-nginx/>

Without using the compound IF workaround, we could simplify the last two rules to OPTIONS or !OPTIONS.  This means that for an options request without a Origin: header, it would send three extra headers to the client.  And for a GET/PUT/POST, etc, it would send one extra header.

@cherrycl do you think the simplified IF logic would work, or do you think we should do the complex IF logic?

**Comment by cloudxxx8** [comment]
@bnevis-i we think it is fine to simplify the rule in nginx.  the actual CORS logic is still located in the individual service.  thanks for the suggestion.

**Comment by bnevis-i** [comment]
@cherrycl What's your take.  Our choices are

- 6 or 4 extra headers (current implementation, simplest)
- 3 or 1 extra headers (simple logic tweak)
- No extra headers (very complex logic tweak)

**Comment by cloudxxx8** [comment]
> @cloudxxx8 Actually, it's an either/or choice -- enable it in the service, or enable it at the API gateway. The API gateway's auth check will block access before it gets sent to the backend; it only makes sense to do CORS in the service if you are not using EdgeX's gateway. A command line client can spoof the CORS headers, so all the protection logic for CORS (excluding token auth) is actually in the browser.

you are right. that makes sense.

**Comment by cherrycl** [comment]
@bnevis-i Thanks for the summary.
I read through the article, and it just states:

> Note that in any access control request, the [Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin) header is always sent.

However, there is no statement about the headers should not be sent.
As a result, I believe the simplest option is enough.  There is no harm to send back more headers.

- 6 or 4 extra headers (current implementation, simplest)

If you agree with that, we can close this issue now.

**Comment by cloudxxx8** [comment]
@bnevis-i Let's use the simplest option.  Is it ok?

- 6 or 4 extra headers (current implementation, simplest)

**Comment by bnevis-i** [comment]
@cloudxxx8   This is what is implemented.  Closing the request.
